Add initial explainer for openable API. (#1149)
openui/open-ui: site/src/pages/components/openable.explainer.mdx

```javascript
function supportsOpenable() {
  return HTMLElement.prototype.hasOwnProperty('openable')
}
```

## Events

An event is fired synchronously when an openable is open or close. This event can be used, for example, to populate content for the openable just in time before it is opened, or update server data when it closes. The event provides a `currentState` and `newState` for the openable. The value of these properties can either be "open" or "closed". The events looks like this:

```javascript
openable.addEventListener('beforetoggle', (beforeToggleEvent) => {
  if (beforeToggleEvent.currentState === 'closed') console.info('Openable is closed')
  if (beforeToggleEvent.newState === 'open') console.info('Openable is being shown')
})
```

The `beforetoggle` event is cancellable when the `newState` is equal to "open". Doing so keeps the openable from being opened. You can't cancel the closing of an openable. The `beforetoggle` event is fired synchronously.

Additionally an asynchronous non-cancellable `toggle` event is fired.

## Focus Management

TODO

## Accessibility / Semantics

> Since the `openable` content attribute can be applied to any element, and this only impacts the element's presentation, this addition does not have any direct semantic or accessibility impact. The element with the `openable` attribute will keep its existing semantics and AOM representation. For example, `<article openable>...</article>` will continue to be exposed as an implicit `role=article`. Similarly, ARIA can be used to modify accessibility mappings in [the normal way](https://w3c.github.io/html-aria/), for example `<div openable role=note>...</div>`.

Reviewer: ```suggestion
The `openable` content attribute can be applied to any element, where in the hidden or "collapsed" state, the element will be hidden from the accessibility tree.  Upon invoking. the element will become available both visually, and via the browser's accessibility tree.  Beyond this behavior, the element with the `openable` attribute will otherwise keep its existing semantics and AOM representation. For example, `<article openable>...</article>` will continue to be exposed as an implicit `role=article`. Similarly, ARIA can be used to modify accessibility mappings [per the allowances defined in the ARIA in HTML specification](https://w3c.github.io/html-aria/).  For example `<div openable role=note>...</div>`.
```

As mentioned in the [Declarative Triggers](#declarative-triggers) section, accessibility mappings will be automatically configured to associate the openable with its trigger element, as needed.

## Disallowed elements

TODO

# Example Use Cases

TODO

# The Choices Made in this API

Many decisions and choices were made in the design of this API, and those decisions were made via numerous discussions (live and on issues) in [OpenUI](https://open-ui.org/), a WHATWG [Community Group](https://open-ui.org/working-mode).

## Alternatives Considered

> TODO

Reviewer: ```suggestion
TODO

An alternate to the `defaultopen` attribute could be to allow the `openable` attribute to declare its initial state.  E.g., `<div openable="open">` or similar.  
```

## Why a content attribute?

> TODO

Reviewer: ```suggestion
Similar to some of the reasons why `popover` evolved into an attribute, many types of elements may need to be shown/hidden or "openable" - with no one single element that could represent all scenarios.  

As an attribute, a controlling element (button) can toggle the openable state of a list of hyperlinks, a table row, or another section or fieldset of related content, etc.  A content attribute allows this <em>behavior</em> to be applied to all of these use cases, and more - allowing the flexibility of developers to use the attribute on <em>any element</em>.

Here, a developer is building a table where additional rows of content can be shown or hidden by use of command buttons. 
```html
<table>
  ...
  <tr>
    <th>Name of row</th>
    <td>...</td>
    <td>
       <button commandfor="moar" command="toggle-openable">view additional rows</button>
    </td>
  </tr>
  <tr openable id="moar">...</tr>
  ...
```

In the following example, a developer is creating a sidebar navigation where each link has a sibling command button to toggle the visibility of sub-navigation items:

```html
<nav aria-label="secondary">
  ...
  <ul>
    ...
    <li>
      <a href="products.html>Our Products</a>
      <button commandfor="subducts" 
        command="toggle-openable" 
        aria-label="our products sub-navigation">
       <!-- svg chevron goes here -->
      </button>
      <ul>
        <li><a href=...>...</a>
        <li><a href=...>...</a>
        ...
      </ul>
    </li>
     ...
  </ul>
</nav>
```

Author: I think the formatting messed up so some of this isn't in the commit suggestion.

## Accessibility / Semantics

The `openable` content attribute can be applied to any element, where in the hidden or "collapsed" state, the element will be hidden from the accessibility tree.  Upon invoking. the element will become available both visually, and via the browser's accessibility tree.  Beyond this behavior, the element with the `openable` attribute will otherwise keep its existing semantics and AOM representation. For example, `<article openable>...</article>` will continue to be exposed as an implicit `role=article`. Similarly, ARIA can be used to modify accessibility mappings [per the allowances defined in the ARIA in HTML specification](https://w3c.github.io/html-aria/).  For example `<div openable role=note>...</div>`.

As mentioned in the [Declarative Triggers](#declarative-triggers) section, accessibility mappings will be automatically configured to associate the openable with its trigger element, as needed.

## Disallowed elements

TODO

# Example Use Cases

> TODO

Reviewer: every time i try to add this in as a suggestion, the suggestion breaks and the second example is left out / treated as a comment following the suggestion.  so, i'm just going to post both here as a comment:

---

Here, a developer is building a table where additional rows of content can be shown or hidden by use of command buttons. 
```html
<table>
  ...
  <tr>
    <th>Name of row</th>
    <td>...</td>
    <td>
       <button commandfor="moar" command="toggle-openable">view additional rows</button>
    </td>
  </tr>
  <tr openable id="moar">...</tr>
  ...
</table>
```
In the following example, a developer is creating a sidebar navigation where each link has a sibling command button to toggle the visibility of sub-navigation items:

```html
<nav aria-label="secondary">
  ...
  <ul>
    ...
    <li>
      <a href="products.html>Our Products</a>
      <button commandfor="subducts" 
        command="toggle-openable" 
        aria-label="our products sub-navigation">
       <!-- svg chevron goes here -->
      </button>
      <ul>
        <li><a href=...>...</a>
        <li><a href=...>...</a>
        ...
      </ul>
    </li>
     ...
  </ul>
</nav>
```

Author: I've uploaded this manually to that section .

# The Choices Made in this API

Many decisions and choices were made in the design of this API, and those decisions were made via numerous discussions (live and on issues) in [OpenUI](https://open-ui.org/), a WHATWG [Community Group](https://open-ui.org/working-mode).

## Alternatives Considered

TODO

An alternate to the `defaultopen` attribute could be to allow the `openable` attribute to declare its initial state.  E.g., `<div openable="open">` or similar.  

## Why a content attribute?

> TODO

Reviewer: ```suggestion
Similar to some of the reasons why `popover` evolved into an attribute, many types of elements may need to be shown/hidden or "openable" - with no one single element that could represent all scenarios.  

As an attribute, a controlling element (button) can toggle the openable state of a list of hyperlinks, a table row, or another section or fieldset of related content, etc.  A content attribute allows this <em>behavior</em> to be applied to all of these use cases, and more - allowing the flexibility of developers to use the attribute on <em>any element</em>.
```

## Alternatives Considered

TODO

An alternate to the `defaultopen` attribute could be to allow the `openable` attribute to declare its initial state.  E.g., `<div openable="open">` or similar.  

## Why a content attribute?

Similar to some of the reasons why `popover` evolved into an attribute, many types of elements may need to be shown/hidden or "openable" - with no one single element that could represent all scenarios.  

As an attribute, a controlling element (button) can toggle the openable state of a list of hyperlinks, a table row, or another section or fieldset of related content, etc.  A content attribute allows this <em>behavior</em> to be applied to all of these use cases, and more - allowing the flexibility of developers to use the attribute on <em>any element</em>.

## Design decisions (via [OpenUI](https://open-ui.org/))

TODO

## Questions

Do we need a source property in an options bag for the open and toggle methods like popover has?

> How do we ensure the command button has its aria-expanded state updated in the find-in-page matching scenario?

Author: ```suggestion
```
